File: py/aoc2021/day7.py
import math


def part1(lines):
    """
    >>> part1(["16,1,2,0,4,2,7,1,2,14"])
    37
    """
    nums = sorted(int(word) for line in lines for word in line.strip().split(","))
    median = nums[len(nums) // 2]
    return sum(abs(num - median) for num in nums)


def t(n):
    return n * (n + 1) // 2


def part2(lines):
    """
    >>> part2(["16,1,2,0,4,2,7,1,2,14"])
    168
    """
    nums = list(int(word) for line in lines for word in line.strip().split(","))
    mean = sum(nums) / len(nums)
    mean0 = math.floor(mean)
    mean1 = math.ceil(mean)
    return min(
        sum(t(abs(num - mean0)) for num in nums),
        sum(t(abs(num - mean1)) for num in nums),
    )


parts = (part1, part2)

File: kt/src/commonMain/kotlin/com/github/ephemient/aoc2021/Day7.kt
package com.github.ephemient.aoc2021

import kotlin.math.abs
import kotlin.math.roundToInt

class Day7(lines: List<String>) {
    private val nums = lines.flatMap { line -> line.splitToSequence(',').map { it.toInt() } }

    fun part1(): Int {
        val median = nums.sorted()[nums.size / 2]
        return nums.sumOf { abs(it - median) }
    }

    fun part2(): Int {
        val mean = nums.sum().toDouble() / nums.size
        val mean0 = mean.roundToInt()
        val mean1 = if (mean0 <= mean) mean0 + 1 else mean0 - 1
        return minOf(nums.sumOf { t(abs(it - mean0)) }, nums.sumOf { t(abs(it - mean1)) })
    }

    companion object {
        private fun t(x: Int): Int = x * (x + 1) / 2
    }
}
